bin/resistance_marker_module.py

```diff
@@ -3,379 +3,480 @@
+    """Reads and merges allele data with aligned ASV data."""
+    logging.info(
+        f"Reading allele data from {allele_data_path} and aligned ASV data from {aligned_asv_table_path}.")
+    try:
+        allele_data = pd.read_csv(allele_data_path, sep='\t')
+        aligned_asv_data = pd.read_csv(aligned_asv_table_path, sep='\t')
+        merged_data = allele_data.merge(
+            aligned_asv_data, left_on=['SampleID', 'Locus', 'ASV'],
+            right_on=['sampleID', 'refid', 'asv'], how='left'
+        )
+        logging.info("Allele data and aligned ASV data merged successfully.")
+        return merged_data
+    except Exception as e:
+        logging.error(f"Failed to read or merge data: {e}")
+        raise
 
-    logging.debug(f"Entering `parse_pseudo_cigar` with string={string}, orientation={orientation}, refseq_len={refseq_len}")
 
-    if not isinstance(string, str):
-        logging.error(f"Expected a string for `parse_pseudo_cigar` but got {type(string)}. Value: {string}")
-        raise TypeError(f"Expected a string for `parse_pseudo_cigar` but got {type(string)}. Value: {string}")
+class ProcessMarkerInfo:
+    def __init__(self, ref_sequences):
+        self.ref_sequences = ref_sequences
 
-    string = string.replace("=", "")  # Remove the "=" separator
-    logging.debug(f"Removed '=' separator. Modified string={string}")
+    def extract_marker_info(self, row):
+        """Adds marker start position relative to strand and reference to a row of marker information."""
+        logging.debug(
+            f"Extracting marker information for row with Locus {row['Locus']}")
+
+        locus = row['Locus']
+        strand = row['strand']
+        codon_start = row['CodonStart']
+        ref_seq = self.ref_sequences[locus].seq
+
+        if strand == '-':
+            refseq_len = len(ref_seq)
+            ref_seq = ref_seq.reverse_complement()
+            relative_codon_start = refseq_len - codon_start - 2
+        else:
+            relative_codon_start = codon_start - 1
+
+        codon_end = relative_codon_start + 3
+        ref_codon = str(ref_seq[relative_codon_start: codon_end])
+        return {
+            'GeneID': row.GeneID,
+            'Gene': row.Gene,
+            'CodonID': row.CodonID,
+            'strand': strand,
+            'Locus': locus,
+            'CodonStart': codon_start,
+            'relativeCodonStart': relative_codon_start,
+            'CodonEnd': codon_end,
+            'RefCodon': ref_codon
+        }
+
+    def extract_position_and_reference_for_markers(self, res_marker_table):
+        """Extracts position and reference information for markers."""
+        logging.info(
+            f"Extracting position and reference information for {len(res_marker_table)} markers.")
+        marker_info = res_marker_table.apply(self.extract_marker_info, axis=1)
+        res_marker_table = pd.DataFrame(marker_info.tolist())
+        res_marker_table['RefAA'] = res_marker_table['RefCodon'].apply(
```

Comment:
If there are any indels in the codon then we say we can't classify the marker and the codon and refcodon will be 'X', this line of code is just to only translate the ones which are classified

Reply:
I see, would it be good to have a have sanity check to make sure the RefCodon always translates in case the end user makes a modification? It's probably not incredibly important and there may be other checks downstream for None in that column

Reply:
If the refCodon doesn't translate then it would be `*`. I'm not sure what you mean by the end sure making a modification?

Reply:
Nevermind, I missed that the ref code was being extracted from the reference sequence still

Reply:
I think N's were my main original concern, which may exist if the user creates a reference sequence, albeit probably unlikely

Reply:
If there was an N in the reference codon then it would be in the refCodon column e.g. ANT and then translate would handle it for the amino acid. If it couldn't translate it it would be 'X'. Do you think it should be handled another way?

Reply:
I think that the X is fine if there is an N, we may want to catch the exception that is thrown if the codon is invalid and handle it. Also, if X is returned by translate(), this line `'X' if not translate(x) else translate(x)` can be reduced to `translate(x)`.

Reply:
I would leave it as is as a stand alone function. We're currently handling indels by saying that they are unclassified so 'X' goes into this line of code. But if we were to change how to handle that then this would break as it can't translate 'AC-' for example

Reply:
Okay, so there are 1.5 problems here, the exception and the line of code.

The exception is thrown when an indel is seen as it can't be translated and is currently not handled - I think that we should be handling exceptions where possible, but given that this should be rare, I think this can just be a note in the documentation. Now if pandas is handling the exception somehow in the background, that would be good to know and may make  'X' if not translate(x) else translate(x)` a reasonable approach. If this is not the case, then please see my second point below.

For the code,  'X' if not translate(x) else translate(x)`, translate is returning the codon or 'X' if it cannot be translated. If there is an indel, then it will throw an exception. I don't see why we have `not translate(x)` if we do not need to worry about a null return. 

The source of my confusion is `not translate(x)`, which does not seem to be needed given the documentation. I do not see an example of this in your test cases as well.

Reply:
Sorry, I've looked at why I did this properly now. So if we have codons with indels then we say they're undetermined and convert them to 'X'. When translated they're null and we want them to be 'X', hence handling it like this.

Reply:
Okay, thanks that's good to know. I don't quite see where the 'X' assignment happens before this lambda is called, but the line makes more sense now